test/addressBook.test.ts

```diff
@@ -0,0 +1,305 @@
+import {
+  MockERC20Instance,
+  WhitelistInstance,
+  OtokenFactoryInstance,
+  UpgradeableContractV1Instance,
+  UpgradeableContractV2Instance,
+  OwnedUpgradeabilityProxyInstance,
+  AddressBookInstance,
+} from '../build/types/truffle-types'
+
+const {expectRevert} = require('@openzeppelin/test-helpers')
+
+const MockERC20 = artifacts.require('MockERC20.sol')
+const OtokenFactory = artifacts.require('OtokenFactory.sol')
+const MarginPool = artifacts.require('MarginPool.sol')
+const UpgradeableContractV1 = artifacts.require('UpgradeableContractV1.sol')
+const UpgradeableContractV2 = artifacts.require('UpgradeableContractV2.sol')
+const OwnedUpgradeabilityProxy = artifacts.require('OwnedUpgradeabilityProxy.sol')
+const AddressBook = artifacts.require('AddressBook.sol')
+const Whitelist = artifacts.require('Whitelist.sol')
+// address(0)
+
+contract('AddressBook', ([owner, otokenImplAdd, marginPoolAdd, random]) => {
+  // ERC20 mocks
+  let weth: MockERC20Instance
+  // addressbook instance
+  let addressBook: AddressBookInstance
+
+  before('Deployment', async () => {
+    // deploy WETH token
+    weth = await MockERC20.new('WETH', 'WETH')
+    // deploy AddressBook token
+    addressBook = await AddressBook.new()
+  })
+
+  describe('Set otoken implementation address', () => {
+    it('should revert adding otoken implementation address from non-owner address', async () => {
+      await expectRevert(addressBook.setOtokenImpl(otokenImplAdd, {from: random}), 'Ownable: caller is not the owner')
+    })
+
```

Comment:
for oTokens, the owner is responsible for ensuring that a valid address is put in? The contract won't fail if I put in the a random addrres?

Reply:
yes, for all addresses, owner is responsible for making sure that `address X` is related to that contract/module.
The Owner in the beginning is the Opyn team through the Timelock contract, later will be a governance module.